src/main/java/bio/terra/service/duos/DuosService.java

```diff
@@ -0,0 +1,76 @@
+package bio.terra.service.duos;
+
+import bio.terra.model.DuosFirecloudGroupModel;
+import bio.terra.service.auth.iam.IamService;
+import bio.terra.service.auth.iam.exception.IamConflictException;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Optional;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DuosService {
+  private static final Logger logger = LoggerFactory.getLogger(DuosService.class);
+
+  private final DuosDao duosDao;
+  private final IamService iamService;
+
+  @Autowired
+  public DuosService(DuosDao duosDao, IamService iamService) {
+    this.duosDao = duosDao;
+    this.iamService = iamService;
+  }
+
+  public Optional<DuosFirecloudGroupModel> retrieveFirecloudGroup(String duosId) {
+    return Optional.ofNullable(duosDao.retrieveFirecloudGroup(duosId));
+  }
+
+  public DuosFirecloudGroupModel createFirecloudGroup(String duosId) {
+    logger.info("Creating Firecloud group for {} users", duosId);
+
+    // First try with the more readable group name.
+    String groupName = constructFirecloudGroupName(duosId);
+    try {
+      iamService.createGroup(groupName);
+    } catch (IamConflictException ex) {
+      logger.warn(
+          "Firecloud group {} already exists: trying creation with a unique name", groupName);
+      groupName = constructUniqueFirecloudGroupName(duosId);
+      iamService.createGroup(groupName);
+    }
+    String groupEmail = iamService.getGroupEmail(groupName);
+    logger.info("Successfully created Firecloud group {} for {} users", groupName, duosId);
+
+    boolean inserted = duosDao.insertFirecloudGroup(duosId, groupName, groupEmail);
```

Comment:
I wonder if to be extra safe, we should put this and the block at line 48 in a try block?

Reply:
That's a good point, especially as we have the fallback behavior to delete the created group if insertion fails.

Reply:
Sharing our Slack convo for posterity:

What I came up with looked suspiciously like something that should be its own step in a flight, given the cleanup behavior:
```
    boolean inserted = false;
    try {
      inserted = duosDao.insertFirecloudGroup(duosId, groupName, groupEmail);
    } catch (Exception ex) {
      // Unexpected exceptions will be thrown after deleting the Firecloud group
      // in the "finally" block.
      throw ex;
    } finally {
      if (!inserted) {
        iamService.deleteGroup(groupName);
        throw new DuosFirecloudGroupInsertException(
            "Firecloud group " + groupName + " was not inserted into the DB and has been deleted");
      }
    }
```
That will be coming soon, in its own ticket: https://broadworkbench.atlassian.net/browse/DR-2787
So at this point, I left a comment to that effect and will lightly refactor to separate group creation from DB insertion when I'm building the flight.

Reply:
And… thanks to @pshapiro4broad 's line of questioning, I was able to streamline this and get our try block after all :)

https://github.com/DataBiosphere/jade-data-repo/pull/1361#discussion_r999929413